File: examples/python-guide/classification_non_Gaussian_data.py
# coding: utf-8
# pylint: disable = invalid-name, C0111
import gpboost as gpb
import numpy as np
import matplotlib.pyplot as plt
from scipy import stats

plt.style.use('ggplot')
print("It is recommended that the examples are run in interactive mode")

"""
Examples of generalized linear Gaussian process and random effects models
for several non-Gaussian likelihoods
"""

# Choose likelihood: either "bernoulli_probit" (=default for binary data), "bernoulli_logit",
#                      "poisson", or "gamma"
likelihood = "bernoulli_probit"

# Non-linear function for simulation
def f1d(x):
    return 1 / (1 + np.exp(-(x - 0.5) * 20)) - 0.5

# Parameters for gpboost in examples below
# Note: the tuning parameters are by no means optimal for all situations considered here
params = {'learning_rate': 0.1, 'min_data_in_leaf': 20, 'objective': likelihood,
          'verbose': 0, 'monotone_constraints': [1, 0]}
num_boost_round = 25
if likelihood in ("bernoulli_probit", "bernoulli_logit"):
    params['objective'] = 'binary'

# --------------------Grouped random effects model----------------
# Simulate data
n = 5000  # number of samples
m = 500  # number of groups
np.random.seed(1)
# simulate grouped random effects
group = np.arange(n)  # grouping variable
for i in range(m):
    group[int(i * n / m):int((i + 1) * n / m)] = i
b1 = 0.5 * np.random.normal(size=m)  # simulate random effects
eps = b1[group]
eps = eps - np.mean(eps)
# simulate fixed effects
X = np.random.rand(n, 2)
f = f1d(X[:, 0])
# simulate response variable
if likelihood == "bernoulli_probit":
    probs = stats.norm.cdf(f + eps)
    y = np.random.uniform(size=n) < probs
    y = y.astype(np.float64)
elif likelihood == "bernoulli_logit":
    probs = 1 / (1 + np.exp(-(f + eps)))
    y = np.random.uniform(size=n) < probs
    y = y.astype(np.float64)
elif likelihood == "poisson":
    mu = np.exp(f + eps)
    y = stats.poisson.ppf(np.random.uniform(size=n), mu=mu)
elif likelihood == "gamma":
    mu = np.exp(f + eps)
    y = mu * stats.gamma.ppf(np.random.uniform(size=n), a=1)
fig1, ax1 = plt.subplots()
ax1.hist(y, bins=50)  # visualize response variable

# create dataset for gpb.train
data_train = gpb.Dataset(X, y)
# Train model
gp_model = gpb.GPModel(group_data=group, likelihood=likelihood)
# Use the option "trace": true to monitor convergence of hyperparameter estimation of the gp_model. E.g.:
# gp_model.set_optim_params(params={"trace": True})
bst = gpb.train(params=params,
                train_set=data_train,
                gp_model=gp_model,
                num_boost_round=num_boost_round)
gp_model.summary()  # Trained random effects model

# Make predictions
nplot = 200  # number of predictions
X_test_plot = np.column_stack((np.linspace(0, 1, nplot), np.zeros(nplot)))
group_data_pred = -np.ones(nplot)
# Predict response variable
pred_resp = bst.predict(data=X_test_plot, group_data_pred=group_data_pred, raw_score=False)
# Predict latent variable including variance
pred = bst.predict(data=X_test_plot, group_data_pred=group_data_pred,
                   predict_var=True, raw_score=True)

# Visualize predictions
fig1, ax1 = plt.subplots()
ax1.plot(X_test_plot[:, 0], f1d(X_test_plot[:, 0]), linewidth=2, label="True F")
ax1.plot(X_test_plot[:, 0], pred['fixed_effect'], linewidth=2, label="Pred F")
if likelihood in ("bernoulli_probit", "bernoulli_logit"):
    ax1.scatter(X[:, 0], y, linewidth=2, color="black", alpha=0.02)
ax1.set_title("Data, true and predicted latent function F")
ax1.legend()

fig1, ax1 = plt.subplots()
ax1.plot(X_test_plot[:, 0], pred_resp['response_mean'], linewidth=2, label="Pred response")
ax1.scatter(X[:, 0], y, linewidth=2, color="black", alpha=0.02)
ax1.set_title("Data and predicted response variable")
ax1.legend()

# Cross-validation for finding number of iterations
gp_model = gpb.GPModel(group_data=group, likelihood=likelihood)
cvbst = gpb.cv(params=params, train_set=data_train,
               gp_model=gp_model, use_gp_model_for_validation=True,
               num_boost_round=200, early_stopping_rounds=5,
               nfold=4, verbose_eval=True, show_stdv=False, seed=1)
print("Best number of iterations: " + str(np.argmin(next(iter(cvbst.values())))))

# Showing training loss
print('Showing training loss...')
gp_model = gpb.GPModel(group_data=group, likelihood=likelihood)
bst = gpb.train(params=params,
                train_set=data_train,
                gp_model=gp_model,
                num_boost_round=5,
                valid_sets=data_train)


# --------------------Gaussian process model----------------
# Simulate data
ntrain = 500  # number of samples
np.random.seed(4)
# training and test locations (=features) for Gaussian process
coords_train = np.column_stack((np.random.uniform(size=ntrain), np.random.uniform(size=ntrain)))
# exclude upper right corner
excl = ((coords_train[:, 0] >= 0.7) & (coords_train[:, 1] >= 0.7))
coords_train = coords_train[~excl, :]
ntrain = coords_train.shape[0]
nx = 30  # test data: number of grid points on each axis
coords_test_aux = np.arange(0, 1, 1 / nx)
coords_test_x1, coords_test_x2 = np.meshgrid(coords_test_aux, coords_test_aux)
coords_test = np.column_stack((coords_test_x1.flatten(), coords_test_x2.flatten()))
coords = np.row_stack((coords_train, coords_test))
ntest = nx * nx
n = ntrain + ntest
# Simulate fixed effects
X_train = np.random.rand(ntrain, 2)
X_test = np.column_stack((np.linspace(0, 1, ntest), np.zeros(ntest)))
X = np.row_stack((X_train, X_test))
f = f1d(X[:, 0])
# Simulate spatial Gaussian process
sigma2_1 = 0.25  # marginal variance of GP
rho = 0.1  # range parameter
D = np.zeros((n, n))  # distance matrix
for i in range(0, n):
    for j in range(i + 1, n):
        D[i, j] = np.linalg.norm(coords[i, :] - coords[j, :])
        D[j, i] = D[i, j]
Sigma = sigma2_1 * np.exp(-D / rho) + np.diag(np.zeros(n) + 1e-20)
C = np.linalg.cholesky(Sigma)
b1 = np.random.normal(size=n)  # simulate random effects
eps = C.dot(b1)
eps = eps - np.mean(eps)
# simulate response variable
if likelihood == "bernoulli_probit":
    probs = stats.norm.cdf(f + eps)
    y = np.random.uniform(size=n) < probs
    y = y.astype(np.float64)
elif likelihood == "bernoulli_logit":
    probs = 1 / (1 + np.exp(-(f + eps)))
    y = np.random.uniform(size=n) < probs
    y = y.astype(np.float64)
elif likelihood == "poisson":
    mu = np.exp(f + eps)
    y = stats.poisson.ppf(np.random.uniform(size=n), mu=mu)
elif likelihood == "gamma":
    mu = np.exp(f + eps)
    y = mu * stats.gamma.ppf(np.random.uniform(size=n), a=1)
# Split into training and test data
y_train = y[0:ntrain]
data_train = gpb.Dataset(X_train, y_train)  # create dataset for gpb.train
y_test = y[ntrain:n]
eps_test = eps[ntrain:n]
plt.hist(y_train, bins=50)  # visualize response variable

# Train model
gp_model = gpb.GPModel(gp_coords=coords_train, cov_function="exponential", likelihood=likelihood)
# Use the option "trace": true to monitor convergence of hyperparameter estimation of the gp_model. E.g.:
# gp_model.set_optim_params(params={"trace": True})
print("Train GPBoost model with GP model (takes a few seconds)")
bst = gpb.train(params=params,
                train_set=data_train,
                gp_model=gp_model,
                num_boost_round=num_boost_round)
gp_model.summary()  # Trained GP model

# Predict response variable
pred_resp = bst.predict(data=X_test, gp_coords_pred=coords_test, raw_score=False)
# Predict latent variable including variance
pred = bst.predict(data=X_test, gp_coords_pred=coords_test, predict_var=True, raw_score=True)

if likelihood in ("bernoulli_probit", "bernoulli_logit"):
    print("Test error:")
    pred_binary = pred_resp['response_mean'] > 0.5
    pred_binary = pred_binary.astype(np.float64)
    print(np.mean(pred_binary != y_test))
else:
    print("Test root mean square error:")
    print(np.sqrt(np.mean((pred_resp['response_mean'] - y_test) ** 2)))
print("Test root mean square error for latent GP:")
print(np.sqrt(np.mean((pred['random_effect_mean'] - eps_test) ** 2)))

# Visualize predictions and compare to true values
fig, axs = plt.subplots(2, 2)
# data and true GP
eps_test_plot = eps_test.reshape((nx, nx))
CS = axs[0, 0].contourf(coords_test_x1, coords_test_x2, eps_test_plot)
axs[0, 0].plot(coords_train[:, 0], coords_train[:, 1], '+', color="white")
axs[0, 0].set_title("True latent GP and training locations")
# predicted latent GP mean
pred_mu_plot = pred['random_effect_mean'].reshape((nx, nx))
CS = axs[0, 1].contourf(coords_test_x1, coords_test_x2, pred_mu_plot)
axs[0, 1].set_title("Predicted latent GP mean")
# prediction uncertainty
pred_var_plot = pred['random_effect_cov'].reshape((nx, nx))
CS = axs[1, 0].contourf(coords_test_x1, coords_test_x2, pred_var_plot)
axs[1, 0].set_title("Predicted latent GP standard deviation")
# latent predictor function F
axs[1, 1].plot(X_test[:, 0], f1d(X_test[:, 0]), linewidth=2, label="True F")
axs[1, 1].plot(X_test[:, 0], pred['fixed_effect'], linewidth=2, label="Pred F")
axs[1, 1].set_title("Predicted and true F")
axs[1, 1].legend()

# Cross-validation for finding number of iterations (takes a few seconds)
gp_model = gpb.GPModel(gp_coords=coords_train, cov_function="exponential", likelihood=likelihood)
cvbst = gpb.cv(params=params, train_set=data_train,
               gp_model=gp_model, use_gp_model_for_validation=True,
               num_boost_round=200, early_stopping_rounds=5,
               nfold=4, verbose_eval=True, show_stdv=False, seed=1)
print("Best number of iterations: " + str(np.argmin(next(iter(cvbst.values())))))

File: R-package/demo/classification_non_Gaussian_data.R
## Examples for combining tree-boosting with Gaussian process and random effects models
##    for several non-Gaussian likelihoods
## See the examples in GPBoost_algorithm.R for more functionality
## Author: Fabio Sigrist

library(gpboost)

## Choose likelihood: either "bernoulli_probit" (=default for binary data), "bernoulli_logit",
##                       "poisson", or "gamma"
likelihood <- "bernoulli_probit"

# Non-linear prior mean function for simulation in examples below
f1d <- function(x) 1/(1+exp(-(x-0.5)*10)) - 0.5
sim_non_lin_f <- function(n){
  X <- matrix(runif(2*n),ncol=2)
  f <- f1d(X[,1])
  return(list(X=X,f=f))
}

# Parameters for gpboost in examples below 
# Note: the tuning parameters are by no means optimal for all situations considered here
params <- list(learning_rate = 0.1, min_data_in_leaf = 20,
               objective = likelihood, monotone_constraints = c(1,0))
nrounds <- 25
if (likelihood=="bernoulli_logit") nrounds <- 50
if (likelihood %in% c("bernoulli_probit","bernoulli_logit")) params$objective="binary"

#--------------------Combine tree-boosting and grouped random effects model----------------
# Simulate data
n <- 5000 # number of samples
m <- 500 # number of groups
set.seed(1)
# Simulate random and fixed effects
group <- rep(1,n) # grouping variable
for(i in 1:m) group[((i-1)*n/m+1):(i*n/m)] <- i
b1 <- sqrt(0.5) * rnorm(m)
eps <- b1[group]
eps <- eps - mean(eps)
sim_data <- sim_non_lin_f(n=n)
f <- sim_data$f
X <- sim_data$X
# Simulate response variable
if (likelihood == "bernoulli_probit") {
  probs <- pnorm(f+eps)
  y <- as.numeric(runif(n) < probs)
} else if (likelihood == "bernoulli_logit") {
  probs <- 1/(1+exp(-(f+eps)))
  y <- as.numeric(runif(n) < probs)
} else if (likelihood == "poisson") {
  mu <- exp(f+eps)
  y <- qpois(runif(n), lambda = mu)
} else if (likelihood == "gamma") {
  mu <- exp(f+eps)
  y <- qgamma(runif(n), scale = mu, shape = 1)
}
hist(y,breaks=50)# visualize response variable

#--------------------Training----------------
# Define random effects model
gp_model <- GPModel(group_data = group, likelihood = likelihood)
bst <- gpboost(data = X, label = y, verbose = 0,
               gp_model = gp_model,
               nrounds = nrounds, 
               params = params)
summary(gp_model) # Trained random effects model (true variance = 0.5)

#--------------------Prediction----------------
nplot <- 200# number of predictions
X_test_plot <- cbind(seq(from=0,to=1,length.out=nplot),rep(0.5,nplot))
group_data_pred <- rep(-9999,dim(X_test_plot)[1]) # only new / unobserved groups
# 1. Predict response variable
pred_resp <- predict(bst, data = X_test_plot, group_data_pred = group_data_pred, rawscore = FALSE)
# pred_resp$response_mean contains the (mean) predictions of the response variable
#   which combines predictions from the tree ensemble and the random effects
# pred_resp$response_var contains the predictive variances (if predict_var=TRUE)
# 2. Predict latent variable and variance
pred <- predict(bst, data = X_test_plot, group_data_pred = group_data_pred,
                predict_var=TRUE, rawscore = TRUE)
# pred_resp$fixed_effect contains the predictions for the latent fixed effects / tree ensemble
# pred_resp$random_effect_mean contains the mean predictions for the latent random effects
# pred_resp$random_effect_cov contains the predictive (co-)variances (if predict_var=TRUE) of the random effects

# Visualize predictions
x <- seq(from=0,to=1,length.out=200)
plot(x,f1d(x),type="l",lwd=3,col=2,main="Data, true and predicted latent function F")
points(X[,1],y,col=rgb(0,0,0,alpha=0.1))
lines(X_test_plot[,1],pred$fixed_effect,col=4,lwd=3)
legend(legend=c("True F","Pred F"),"bottomright",bty="n",lwd=3,col=c(2,4))

plot(X[,1],y,col=rgb(0,0,0,alpha=0.1),main="Data and predicted response variable")
lines(X_test_plot[,1],pred_resp$response_mean,col=3,lwd=3)

#--------------------Choosing tuning parameters----------------
param_grid = list("learning_rate" = c(1,0.1,0.01), "min_data_in_leaf" = c(1,10,100),
                  "max_depth" = c(1,3,5,10))
gp_model <- GPModel(group_data = group, likelihood = likelihood)
dataset <- gpb.Dataset(data = X, label = y)
set.seed(10)
opt_params <- gpb.grid.search.tune.parameters(param_grid = param_grid,
                                              params = params,
                                              num_try_random = NULL,
                                              nfold = 4,
                                              data = dataset,
                                              gp_model = gp_model,
                                              verbose_eval = 1,
                                              nrounds = 1000,
                                              early_stopping_rounds = 10)
print(paste0("Best parameters: ",paste0(unlist(lapply(seq_along(opt_params$best_params), function(y, n, i) { paste0(n[[i]],": ", y[[i]]) }, y=opt_params$best_params, n=names(opt_params$best_params))), collapse=", ")))
print(paste0("Best number of iterations: ", opt_params$best_iter))
print(paste0("Best score: ", round(opt_params$best_score, digits=3)))

#--------------------Cross-validation for finding number of iterations----------------
dtrain <- gpb.Dataset(data = X, label = y)
gp_model <- GPModel(group_data = group, likelihood = likelihood)
cvbst <- gpb.cv(params = params,
                data = dtrain,
                gp_model = gp_model,
                nrounds = 200,
                nfold = 4,
                verbose = 1,
                early_stopping_rounds = 5,
                use_gp_model_for_validation = TRUE)
print(paste0("Optimal number of iterations: ", cvbst$best_iter))

#--------------------Using a validation set for finding number of iterations----------------
set.seed(1)
train_ind <- sample.int(n,size=as.integer(0.8*n))
dtrain <- gpb.Dataset(data = X[train_ind,], label = y[train_ind])
dvalid <- gpb.Dataset.create.valid(dtrain, data = X[-train_ind,], label = y[-train_ind])
valids <- list(test = dvalid)
gp_model <- GPModel(group_data = group[train_ind], likelihood = likelihood)
gp_model$set_prediction_data(group_data_pred = group[-train_ind])
bst <- gpb.train(data = dtrain,
                 gp_model = gp_model,
                 nrounds = 100,
                 params = params,
                 verbose = 1,
                 valids = valids,
                 early_stopping_rounds = 5,
                 use_gp_model_for_validation = TRUE)
print(paste0("Optimal number of iterations: ", bst$best_iter,
             ", best test error: ", bst$best_score))


#--------------------Compare to generalized linear mixed effects model----------------
X_lin <- cbind(rep(1,n),X)# Add intercept column
gp_model <- fitGPModel(group_data = group, likelihood = likelihood, y = y, X = X_lin)
summary(gp_model)

X_test_lin <- cbind(rep(1,dim(X_test_plot)[1]),X_test_plot)
# Predict latent variable
pred_lin <- predict(gp_model, y = y, X_pred = X_test_lin,
                    group_data_pred = group_data_pred, predict_response = FALSE)
# Predict response variable
pred_lin_resp <- predict(gp_model, y = y, X_pred = X_test_lin,
                         group_data_pred = group_data_pred, predict_response = TRUE)

# Plot results
plot(x,f1d(x),type="l",lwd=3,col=2,main="Data, true and fitted function")
points(X[,1],y,col=rgb(0,0,0,alpha=0.1))
lines(X_test_plot[,1],pred$fixed_effect,col=4,lwd=3)
lines(X_test_plot[,1],pred_lin$mu,col=3,lwd=3)
legend(legend=c("True F","Pred F GPBoost","Pred F linear"),"bottomright",bty="n",lwd=3,col=c(2,4,3))


#--------------------Combine tree-boosting and Gaussian process model----------------
# Simulate data
ntrain <- 500
set.seed(1)
# Training and test locations for GP (exlcude upper right square from training locations)
coords_train <- matrix(runif(2)/2,ncol=2)
while (dim(coords_train)[1]<ntrain) {
  coord_i <- runif(2) 
  if (!(coord_i[1]>=0.7 & coord_i[2]>=0.7)) {
    coords_train <- rbind(coords_train,coord_i)
  }
}
nx <- 30 # test data: number of grid points on each axis
x2 <- x1 <- rep((1:nx)/nx,nx)
for(i in 1:nx) x2[((i-1)*nx+1):(i*nx)]=i/nx
coords_test <- cbind(x1,x2)
coords <- rbind(coords_train, coords_test)
ntest <- nx * nx
n <- ntrain + ntest
# Simulate fixed effects
X_train <- matrix(runif(2*ntrain),ncol=2)
x <- seq(from=0,to=1,length.out=nx^2)
X_test <- cbind(x,rep(0,nx^2))
X <- rbind(X_train,X_test)
f <- f1d(X[,1])
# Simulate spatial Gaussian process
sigma2_1 <- 0.25 # marginal variance of GP
rho <- 0.1 # range parameter
D <- as.matrix(dist(coords))
Sigma <- sigma2_1 * exp(-D/rho) + diag(1E-20,n)
C <- t(chol(Sigma))
b_1 <- rnorm(n=n)
eps <- as.vector(C %*% b_1)
eps <- eps - mean(eps)
# Simulate response variable
if (likelihood == "bernoulli_probit") {
  probs <- pnorm(f+eps)
  y <- as.numeric(runif(n) < probs)
} else if (likelihood == "bernoulli_logit") {
  probs <- 1/(1+exp(-(f+eps)))
  y <- as.numeric(runif(n) < probs)
} else if (likelihood == "poisson") {
  mu <- exp(f+eps)
  y <- qpois(runif(n), lambda = mu)
} else if (likelihood == "gamma") {
  mu <- exp(f+eps)
  y <- qgamma(runif(n), scale = mu, shape = 1)
}
# Split into training and test data
y_train <- y[1:ntrain]
dtrain <- gpb.Dataset(data = X_train, label = y_train)
y_test <- y[1:ntest+ntrain]
eps_test <- eps[1:ntest+ntrain]
hist(y_train,breaks=50)# visualize response variable

# Train model
gp_model <- GPModel(gp_coords = coords_train, cov_function = "exponential",
                    likelihood = likelihood)
# Takes a few seconds
bst <- gpb.train(data = dtrain, gp_model = gp_model,
                 nrounds = nrounds, params = params, verbose = 0)
summary(gp_model)# Trained GP model

# Prediction of latent variable
pred <- predict(bst, data = X_test, gp_coords_pred = coords_test,
                predict_var = TRUE, rawscore = TRUE)
# Predict response variable (label)
pred_resp <- predict(bst, data = X_test, gp_coords_pred = coords_test, rawscore = FALSE)
if (likelihood %in% c("bernoulli_probit","bernoulli_logit")) {
  print("Test error:")
  mean(as.numeric(pred_resp$response_mean>0.5) != y_test)
} else {
  print("Test root mean square error:")
  sqrt(mean((pred_resp$response_mean - y_test)^2))
}
print("Test root mean square error for latent GP:")
sqrt(mean((pred$random_effect_mean - eps_test)^2))

# Visualize predictions and compare to true values
library(ggplot2)
library(viridis)
library(gridExtra)
plot1 <- ggplot(data = data.frame(s_1=coords_test[,1],s_2=coords_test[,2],b=eps_test),aes(x=s_1,y=s_2,color=b)) +
  geom_point(size=4, shape=15) + scale_color_viridis(option = "B") + ggtitle("True latent GP and training locations") + 
  geom_point(data = data.frame(s_1=coords_train[,1],s_2=coords_train[,2],y=y_train),aes(x=s_1,y=s_2),size=3, col="white", alpha=1, shape=43)
plot2 <- ggplot(data = data.frame(s_1=coords_test[,1],s_2=coords_test[,2],b=pred$random_effect_mean),aes(x=s_1,y=s_2,color=b)) +
  geom_point(size=4, shape=15) + scale_color_viridis(option = "B") + ggtitle("Predicted latent GP mean")
plot3 <- ggplot(data = data.frame(s_1=coords_test[,1],s_2=coords_test[,2],b=sqrt(pred$random_effect_cov)),aes(x=s_1,y=s_2,color=b)) +
  geom_point(size=4, shape=15) + scale_color_viridis(option = "B") + labs(title="Predicted latent GP standard deviation", subtitle=" = prediction uncertainty")
plot4 <- ggplot(data=data.frame(x=X_test[,1],f=pred$fixed_effect), aes(x=x,y=f)) + geom_line(size=1) +
  geom_line(data=data.frame(x=x,f=f1d(x)), aes(x=x,y=f), size=1.5, color="darkred") +
  ggtitle("Predicted and true F(X)") + xlab("X") + ylab("y")
grid.arrange(plot1, plot2, plot3, plot4, ncol=2)

# Cross-validation for finding number of iterations (takes a few seconds)
gp_model <- GPModel(gp_coords = coords_train, cov_function = "exponential",
                    likelihood = likelihood)
cvbst <- gpb.cv(params = params,
                data = dtrain,
                gp_model = gp_model,
                nrounds = 200,
                nfold = 4,
                verbose = 1,
                early_stopping_rounds = 5,
                use_gp_model_for_validation = TRUE)
print(paste0("Optimal number of iterations: ", cvbst$best_iter))

